### PR TITLE
Add native mesh file I/O (OBJ, STL, PLY)

### PR DESCRIPTION
**Pull request by dhairyashilRG**
## Description

This PR adds native, zero-dependency mesh I/O utilities (`wp.load_mesh()` and `wp.save_mesh()`) to Warp as per [the plan](https://gist.github.com/dhairyashilRG/48d661dcd7162faabaeb2ee4aecc8b51)

Currently, loading basic test meshes in Warp requires either `pxr.UsdGeom` (a massive dependency) or third-party packages like `trimesh`. This forces many basic geometry tests and examples to be gated behind `USD_AVAILABLE`. 

This PR implements lightweight, pure-Python parsers for standard mesh formats (Wavefront OBJ, ASCII/Binary STL, and ASCII/Binary PLY) directly into Warp. This is intended to drop the OpenUSD requirement for testing simple geometries and improve onboarding, not to replace high-performance C++ pipelines for massive assets.

## Checklist

- [x] New and existing tests cover these changes.
- [x] The CHANGELOG.md entry has been added.

## Test plan

- Added a comprehensive suite of unit tests in (`warp/tests/io/test_mesh.py`) for OBJ, STL (ASCII/Binary + vertex deduplication), and PLY formats. 
- Generated standard geometric test assets (`cube.obj`, `cube.stl`, etc.).
- Verified locally on CPU and CUDA (T4, sm_75) via: `uv run --extra dev -m warp.tests -s autodetect -k TestMeshIO`
- 23+ mesh tests pass.

## New feature / enhancement

```python
import warp as wp

# One-liner to load a mesh (format inferred automatically, BVH constructed)
mesh = wp.load_mesh("tests/io/assets/cube.obj")

# Example loading binary STL with simple vertex deduplication
mesh = wp.load_mesh("model.stl", stl_merge_tolerance=1e-5)

# Save back out
wp.save_mesh(mesh, "output.ply")
```

**Out of scope for this PR:** Material/texture loading (.mtl files), animation/sequence loading, and GPU-accelerated parsing (file I/O is CPU-bound).

## Summary by CodeRabbit

* **New Features**
  * Native mesh import/export for OBJ, STL (binary/ASCII), and PLY (binary/ASCII) with zero external dependencies; auto-detection, optional winding flip, and device-aware load/save with round-trip support.

* **Documentation**
  * Changelog updated and a detailed mesh I/O testing/status document added.

* **Tests**
  * New unit and validation tests, deterministic mesh assets, and a validation script covering parsing, formats, edge cases, and round-trips.

* **Examples**
  * Example updated to show alternative mesh-loading via the new I/O.

### CODE REVIEW

File: warp/tests/io/validate_mesh_io.py [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+    print(f"  Generating {num_verts} vertices, {num_tris} triangles...", end=" ")
+
+    # Create vertices on a sphere
+    indices_u = np.arange(num_tris, dtype=np.int32)
+    indices_v = np.arange(num_tris, dtype=np.int32)
+
+    # Simple triangulated plane
+    rng = np.random.default_rng()
+    points = rng.random((num_verts, 3), dtype=np.float32) * 10.0
+    indices = rng.integers(0, num_verts, (num_tris * 3,), dtype=np.int32)
```
[the remark this answers is not in the document]
Comment (dhairyashilRG) [reply warp/tests/io/validate_mesh_io.py]: Removed as per suggeston

File: warp/_src/io/ply.py [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+                            normals[-1][1] = float(values[idx])
+                        elif prop_name == "nz":
+                            normals[-1][2] = float(values[idx])
+                    elif prop_name in ("red", "green", "blue", "alpha"):
+                        if not colors:
+                            colors = [[0, 0, 0] for _ in range(len(points))]
+                        color_idx = {"red": 0, "green": 1, "blue": 2}.get(prop_name)
+                        if color_idx is not None:
+                            colors[-1][color_idx] = int(values[idx])
+                idx += 1
```
[the remark this answers is not in the document]
Comment (dhairyashilRG) [reply warp/_src/io/ply.py]: Resolved as per comment.

File: warp/_src/io/ply.py [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+            f.write(b"property list uchar int vertex_indices\n")
+            f.write(b"end_header\n")
+
+            # Write vertices
+            vertex_format = "<3f"
+            if has_normals:
+                vertex_format += "<3f"
+            if has_colors:
+                vertex_format += "<3B"
+            vertex_struct = struct.Struct(vertex_format)
```
[the remark this answers is not in the document]
Comment (dhairyashilRG) [reply warp/_src/io/ply.py]: Resolved as per comment

File: warp/__init__.py [hunk 4]
```diff
@@ -429,6 +429,13 @@
 from warp._src.paddle import stream_from_paddle as stream_from_paddle
 
 
+# category: Mesh I/O
+
+from warp._src.io.mesh import load_mesh as load_mesh
```
[the remark this answers is not in the document]
Comment (dhairyashilRG) [reply warp/__init__.py:434]: Resolved as per comment

File: warp/_src/io/stl.py [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+            f.read(2)  # attribute byte count
+
+    # Vertex deduplication using spatial hash
+    points, indices = _deduplicate_stl_vertices(raw_points, merge_tolerance)
+
+    return MeshData(
+        points=points,
+        indices=indices,
+        normals=face_normals,
+    )
```
[the remark this answers is not in the document]
Comment (dhairyashilRG) [reply warp/_src/io/stl.py:181]: Now returns MeshData without normals to avoid shape mismatch (stl.py:143-144)

File: warp/tests/io/test_mesh.py [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+
+        # All should have same triangle count
+        self.assertEqual(obj_mesh.indices.shape[0], 3)
+        self.assertEqual(stl_mesh.indices.shape[0], 3)
+        self.assertEqual(ply_mesh.indices.shape[0], 3)
+
+    def test_read_mesh_returns_meshdata(self):
+        """read_mesh() returns MeshData, not wp.Mesh."""
+        from warp._src.io.mesh import MeshData  # noqa: PLC0415
+
```
[the remark this answers is not in the document]
Comment (dhairyashilRG) [reply warp/tests/io/test_mesh.py]: Changed from from warp._src.io.mesh import MeshData to wp.MeshData (test_mesh.py:160-161)

File: warp/tests/io/test_mesh.py [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+            fake_file = os.path.join(tmpdir, "no_extension")
+            with open(fake_file, "w") as f:
+                f.write("v 0 0 0\nv 1 0 0\nv 0 1 0\nf 1 2 3\n")
+
+            mesh = wp.load_mesh(fake_file, format="obj")
+            self.assertEqual(mesh.points.shape[0], 3)
+
+    # ── Winding Order Tests ──
+
+    def test_flip_winding_order(self):
```
[the remark this answers is not in the document]
Comment (dhairyashilRG) [reply warp/tests/io/test_mesh.py:203]: Fixed. Changed format="obj" to file_format="obj" to match the correct parameter name in wp.load_mesh()

File: warp/_src/io/stl.py [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+            raise RuntimeError(f"No triangles found in STL file: '{filename}'")
+
+        # Pre-allocate: worst case = 3 * num_tris unique vertices
+        raw_points = np.empty((num_tris * 3, 3), dtype=np.float32)
+        # STL face normals are read but not used (per-face, not per-vertex)
+        face_normals = np.empty((num_tris, 3), dtype=np.float32)
+
+        for i in range(num_tris):
+            normal = struct.unpack("<3f", f.read(12))
+            face_normals[i] = normal
```
[the remark this answers is not in the document]
Comment (dhairyashilRG) [reply warp/_src/io/stl.py]: Fixed. Removed the unused face_normals allocation. The 12-byte normal per triangle is now simply skipped with f.read(12) instead of allocating memory that was never used.

File: warp/_src/io/ply.py [hunk 9]
```diff
@@ -0,0 +1,419 @@
...
+                break
+            num_verts = struct.unpack("B", count_byte)[0]
+
+            # Read vertex indices (typically int32/uint)
+            idx_format = endian + "i"
+            idx_size = struct.calcsize(idx_format)
+            face_verts = []
+            for _ in range(num_verts):
+                idx = struct.unpack(idx_format, f.read(idx_size))[0]
+                face_verts.append(idx)
```
[the remark this answers is not in the document]
Comment (dhairyashilRG) [reply warp/_src/io/ply.py]: Fixed. The PLY header parser now captures face_index_count_type and face_index_value_type from the header. The binary reader uses a proper PLY-to-struct format mapping instead of hardcoding int32. Supports char, uchar, short, ushort, int, int8, int32, uint, uint8, uint32 types.

File: warp/tests/unittest_suites.py [hunk 10]
```diff
@@ -115,6 +115,7 @@ def default_suite(test_loader: unittest.TestLoader = unittest.defaultTestLoader)
     from warp.tests.interop.test_dlpack import TestDLPack
     from warp.tests.interop.test_jax import TestJax
     from warp.tests.interop.test_torch import TestTorch
+    from warp.tests.io.test_mesh import TestMeshIO, TestMeshIODevices
```
[the remark this answers is not in the document]
Comment (dhairyashilRG) [reply warp/tests/unittest_suites.py:118]: Fixed. Deleted both files as they were not following project conventions (not unittest.TestCase subclasses, hardcoded paths, module-scope code). Their functionality is already covered by existing tests in test_mesh.py: test_mesh_queries_work, test_obj_roundtrip, test_stl_roundtrip, and test_ply_roundtrip.

File: warp/_src/io/stl.py [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+                        indices.extend([vertex_offset, vertex_offset + 1, vertex_offset + 2])
+                        vertex_offset += 3
+
+    if not points:
+        raise RuntimeError(f"No vertices found in STL file: '{filename}'")
+
+    return MeshData(
+        points=np.array(points, dtype=np.float32),
+        indices=np.array(indices, dtype=np.int32),
+    )
```
[the remark this answers is not in the document]
Comment (dhairyashilRG) [reply warp/_src/io/stl.py:244]: This was already fixed in read_stl() at lines 229-232. Both binary and ASCII STL now use the same _deduplicate_stl_vertices() function, ensuring consistent vertex counts and mesh topology.

File: warp/_src/io/ply.py [hunk 12]
```diff
@@ -0,0 +1,431 @@
...
+        if header["has_colors"]:
+            colors = np.empty((vertex_count, 3), dtype=np.uint8)
+
+        # Read vertices
+        vertex_struct = struct.Struct(endian + vertex_format)
+        for i in range(vertex_count):
+            data = vertex_struct.unpack(f.read(vertex_size))
+            # Only access properties that exist in the header
+            points[i] = [data[prop_offsets["x"]], data[prop_offsets["y"]], data[prop_offsets["z"]]]
+
```
[the remark this answers is not in the document]
Comment (dhairyashilRG) [reply warp/_src/io/ply.py:281]: Fixed. Added validation for required vertex position properties (x, y, z) in both _read_ply_binary() and _read_ply_ascii(). The code now checks for these properties before reading vertices and raises a clear RuntimeError with the filename and list of missing properties if any are not found.

File: warp/_src/io/ply.py [hunk 13]
```diff
@@ -0,0 +1,431 @@
...
+                break
+            num_verts = struct.unpack("B", count_byte)[0]
+
+            # Read vertex indices (typically int32/uint)
+            idx_format = endian + "i"
+            idx_size = struct.calcsize(idx_format)
+            face_verts = []
+            for _ in range(num_verts):
+                idx = struct.unpack(idx_format, f.read(idx_size))[0]
+                face_verts.append(idx)
```
[the remark this answers is not in the document]
Comment (dhairyashilRG) [reply warp/_src/io/ply.py]: fixed in commit 0172eba

File: warp/_src/io/ply.py [hunk 14]
```diff
@@ -0,0 +1,482 @@
...
+                if current_element == "vertex" and len(parts) >= 3:
+                    prop_type = parts[1]
+                    prop_name = parts[2]
+                    header_info["vertex_properties"].append((prop_type, prop_name))
+                    if prop_name in ("red", "green", "blue", "alpha"):
+                        header_info["has_colors"] = True
+                elif current_element == "face" and len(parts) >= 5 and parts[1] == "list":
+                    # property list <count_type> <value_type> <name>
+                    header_info["face_index_count_type"] = parts[2]
+                    header_info["face_index_value_type"] = parts[3]
```
[the remark this answers is not in the document]
Comment (dhairyashilRG) [reply warp/_src/io/ply.py]: Added parts[4] in ("vertex_indices", "vertex_index") condition to only capture types for the actual vertex indices property.

File: warp/_src/io/ply.py [hunk 15]
```diff
@@ -0,0 +1,482 @@
...
+                    header_info["vertex_count"] = int(parts[2])
+                elif parts[1] == "face":
+                    header_info["face_count"] = int(parts[2])
+            elif keyword == "property":
+                if current_element == "vertex" and len(parts) >= 3:
+                    prop_type = parts[1]
+                    prop_name = parts[2]
+                    header_info["vertex_properties"].append((prop_type, prop_name))
+                    if prop_name in ("red", "green", "blue", "alpha"):
+                        header_info["has_colors"] = True
```
[the remark this answers is not in the document]
Comment (dhairyashilRG) [reply warp/_src/io/ply.py]: Changed from any() matches to all() matches for both normals (nx, ny, nz) and colors (red, green, blue). Arrays are only allocated when the complete component set exists.

File: warp/_src/io/ply.py [hunk 16]
```diff
@@ -0,0 +1,482 @@
...
+            "face_index_count_type": None,  # Type for vertex count per face
+            "face_index_value_type": None,  # Type for vertex indices
+        }
+
+        current_element = None
+
+        while True:
+            line = f.readline().decode("ascii").strip()
+            if not line:
+                continue
```
[the remark this answers is not in the document]
Comment (dhairyashilRG) [reply warp/_src/io/ply.py:54]: Added EOF checks that raise RuntimeError with clear message in all three header-parsing loops (main header parser, ASCII reader, binary reader).

File: warp/_src/io/stl.py [hunk 17]
```diff
@@ -0,0 +1,332 @@
...
```
[the remark this answers is not in the document]
Comment (dhairyashilRG) [reply warp/_src/io/stl.py:244]: Fixed. Added if in_facet: check before returning MeshData. Also reset facet_vertex_count = 0 after processing endfacet to avoid false positives from properly terminated last facets.

File: warp/_src/io/obj.py [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+                face_verts = []
+                for token in parts[1:]:
+                    # Handle v, v/vt, v//vn, v/vt/vn
+                    v_idx_str = token.split("/")[0]
+                    if not v_idx_str:
+                        continue
+                    v_idx = int(v_idx_str)
+                    # OBJ is 1-indexed; negative = relative to end
+                    if v_idx < 0:
+                        v_idx = len(points) + v_idx + 1
```
[the remark this answers is not in the document]
Comment (dhairyashilRG) [reply warp/_src/io/obj.py]: Fixed. The formula was off by one. Changed from len(points) + v_idx + 1 to len(points) + v_idx. With 3 vertices, -1 now correctly maps to index 2 (last vertex) instead of index 3 (out of bounds). This matches the OBJ specification where negative indices are relative to the end of the vertex list.

File: warp/_src/io/stl.py [hunk 19]
```diff
@@ -0,0 +1,332 @@
...
+
+    Returns:
+        Tuple of (unique_points, unique_indices).
+    """
+    if tolerance <= 0.0:
+        return raw_points, np.arange(len(raw_points), dtype=np.int32)
+
+    # Scale points to integer grid
+    scale = 1.0 / tolerance
+    grid_points = np.round(raw_points * scale).astype(np.int64)
```
[the remark this answers is not in the document]
Comment (dhairyashilRG) [reply warp/_src/io/stl.py:92]: Fixed

File: warp/_src/io/ply.py [hunk 20]
```diff
@@ -0,0 +1,499 @@
...
+        count_format = endian + count_fmt
+        count_size = struct.calcsize(count_format)
+        idx_format = endian + idx_fmt
+        idx_size = struct.calcsize(idx_format)
+
+        # For faces, format is: <count_type> count + <value_type> vertex_indices
+        for _ in range(face_count):
+            count_bytes = f.read(count_size)
+            if not count_bytes:
+                break
```
[the remark this answers is not in the document]
Comment (dhairyashilRG) [reply warp/_src/io/ply.py]: Fixed.

File: warp/_src/io/ply.py [hunk 21]
```diff
@@ -0,0 +1,499 @@
...
+        "int": "i",
+        "int8": "b",
+        "int32": "i",
+        "uint": "I",
+        "uint8": "B",
+        "uint32": "I",
+        "float": "f",
+        "float32": "f",
+        "double": "d",
+    }
```
[the remark this answers is not in the document]
Comment (dhairyashilRG) [reply warp/_src/io/ply.py:226]: Fixed.

File: warp/_src/io/ply.py [hunk 22]
```diff
@@ -0,0 +1,499 @@
...
+                    elif prop_name == "ny" and normals is not None:
+                        normals[vertex_idx][1] = float(values[idx])
+                    elif prop_name == "nz" and normals is not None:
+                        normals[vertex_idx][2] = float(values[idx])
+                    elif prop_name == "red" and colors is not None:
+                        colors[vertex_idx][0] = int(values[idx])
+                    elif prop_name == "green" and colors is not None:
+                        colors[vertex_idx][1] = int(values[idx])
+                    elif prop_name == "blue" and colors is not None:
+                        colors[vertex_idx][2] = int(values[idx])
```
[the remark this answers is not in the document]
Comment (dhairyashilRG) [reply warp/_src/io/ply.py]: Fixed

File: warp/_src/io/obj.py [hunk 23]
```diff
@@ -0,0 +1,170 @@
...
+                        indices.extend([face_verts[0], face_verts[i], face_verts[i + 1]])
+
+    if not points:
+        raise RuntimeError(f"No vertices found in OBJ file: '{filename}'")
+    if not indices:
+        raise RuntimeError(f"No faces found in OBJ file: '{filename}'")
+
+    points_array = np.array(points, dtype=np.float32)
+    indices_array = np.array(indices, dtype=np.int32)
+    normals_array = np.array(normals, dtype=np.float32) if normals else None
```
[the remark this answers is not in the document]
Comment (dhairyashilRG) [reply warp/_src/io/obj.py]: Fixed

File: warp/_src/io/obj.py [hunk 24]
```diff
@@ -0,0 +1,172 @@
...
+        raise RuntimeError(f"No vertices found in OBJ file: '{filename}'")
+    if not indices:
+        raise RuntimeError(f"No faces found in OBJ file: '{filename}'")
+
+    points_array = np.array(points, dtype=np.float32)
+    indices_array = np.array(indices, dtype=np.int32)
+    # Only include normals if count matches vertices (per-vertex normals required)
+    # OBJ files can have different vn count than v count for hard-edge meshes
+    normals_array = np.array(normals, dtype=np.float32) if normals and len(normals) == len(points) else None
+    uvs_array = np.array(uvs, dtype=np.float32) if uvs else None
```
[the remark this answers is not in the document]
Comment (dhairyashilRG) [reply warp/_src/io/obj.py]: Fixed

File: warp/_src/io/obj.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (dhairyashilRG) [reply warp/_src/io/obj.py]: Fixed in commit 5a5e5f5 - normals (and UVs in 9fd571f) are now only included when len(normals) == len(points).

**Comment by dhairyashilRG** [comment]
> ### Greptile Summary
> This PR adds native, zero-dependency mesh I/O (`wp.load_mesh`, `wp.save_mesh`, `wp.read_mesh`) for OBJ, binary/ASCII STL, and binary/ASCII PLY, cleanly structured behind a `MeshData` dataclass and correctly re-exported into the `wp` namespace. Most P1 issues from prior review rounds have been resolved; one new P1 remains:
> 
> * **CHANGELOG.md**: The [Support `indexedarray` fields in `struct` #1327](https://github.com/NVIDIA/warp/issues/1327) `wp.indexedarray`/`@wp.struct` entry was accidentally deleted — almost certainly a rebase conflict artifact. It must be restored before merge.
> 
> ### Confidence Score: 4/5
> Safe to merge after restoring the accidentally deleted CHANGELOG entry for [GH-1327](https://github.com/NVIDIA/warp/issues/1327); all other remaining findings are P2 quality concerns.
> 
> One P1 remains: an unrelated CHANGELOG entry was removed in the diff (likely a rebase artifact). The parsers and public API are well-structured and the prior P1 bugs (PLY normal allocation, struct format string, STL shape mismatch, negative index off-by-one, MeshData not exported) are all resolved. P2 items (PLY unknown-type misalignment, STL triple-parse, write_stl empty-mesh crash) do not block correctness for the intended use cases.
> 
> CHANGELOG.md (deleted [GH-1327](https://github.com/NVIDIA/warp/issues/1327) entry must be restored)
> 
> ### Important Files Changed
> Filename	Overview
> CHANGELOG.md	New mesh I/O entry added correctly; unrelated [GH-1327](https://github.com/NVIDIA/warp/issues/1327) indexedarray entry was accidentally deleted, likely a rebase conflict artifact.
> warp/_src/io/mesh.py	New dispatch layer: MeshData dataclass, load_mesh/read_mesh/save_mesh/flip_winding helpers — clean structure, correct device handling, well-documented.
> warp/_src/io/obj.py	OBJ reader/writer with fan triangulation, negative index fix, normals length guard; UV array still lacks the same length guard as normals.
> warp/_src/io/ply.py	PLY ASCII/binary reader/writer; binary reader silently misaligns data if any vertex property type is not in the internal type_map.
> warp/_src/io/stl.py	STL binary/ASCII reader with spatial-hash deduplication; ASCII format is reparsed up to three times and write_stl crashes with a confusing error on empty meshes.
> warp/tests/io/test_mesh.py	Comprehensive unittest.TestCase-based suite covering OBJ/STL/PLY load, roundtrip, device, winding-order, and error cases; uses proper Path(**file**) relative paths.
> warp/tests/unittest_suites.py	TestMeshIO and TestMeshIODevices correctly added to default_suite as required by AGENTS.md.
> ### Flowchart
> ```mermaid
> %%{init: {'theme': 'neutral'}}%%
> flowchart TD
>     A["wp.load_mesh(filename, ...)"] --> B["read_mesh(filename, ...)"]
>     C["wp.read_mesh(filename, ...)"] --> B
>     B --> D{File exists?\nSize OK?}
>     D -- No --> E[FileNotFoundError / ValueError]
>     D -- Yes --> F{Detect format}
>     F -- ext=obj --> G["read_obj()"]
>     F -- ext=stl --> H["read_stl()"]
>     F -- ext=ply --> I["read_ply()"]
>     H --> J{Binary or ASCII?}
>     J -- binary --> K["_read_binary_stl()"]
>     J -- ascii --> L["_read_ascii_stl()"]
>     K --> M["_deduplicate_stl_vertices()"]
>     L --> N["_deduplicate_stl_vertices()"]
>     I --> O{Format?}
>     O -- ascii --> P["_read_ply_ascii()"]
>     O -- binary --> Q["_read_ply_binary()"]
>     G --> R[MeshData]
>     M --> R
>     N --> R
>     P --> R
>     Q --> R
>     R --> S{flip_winding?}
>     S -- Yes --> T["_apply_flip_winding()"]
>     S -- No --> U[MeshData returned]
>     T --> U
>     U -- load_mesh path --> V["MeshData.to_warp_mesh(device)"]
>     V --> W[wp.Mesh with BVH]
>     X["wp.save_mesh(mesh, filename)"] --> Y{Detect ext}
>     Y -- obj --> Z["write_obj()"]
>     Y -- stl --> AA["write_stl()"]
>     Y -- ply --> AB["write_ply()"]
> ```
> ### Comments Outside Diff (1)
> 1. `CHANGELOG.md`, line 34-36 ([link](https://github.com/nvidia/warp/blob/9fd571f7b7545c24a9a2d0dabeb2757cecf7b3b3/CHANGELOG.md#L34-L36))
>    [P1](#) **Unrelated changelog entry accidentally removed**
>    The [Support `indexedarray` fields in `struct` #1327](https://github.com/NVIDIA/warp/issues/1327) entry for `wp.indexedarray` support in `@wp.struct` was deleted in this diff. This is almost certainly a rebase/merge-conflict resolution artifact — the AGENTS.md guide specifically warns about this pattern ("After rebasing, diff CHANGELOG.md against the target branch to catch duplicate entries or other issues from clean but incorrect merges"). Restoring the dropped lines is needed to preserve the release notes for that feature.
> 
> Reviews (8): Last reviewed commit: ["fix(io): Discard OBJ UVs when count mism..."](https://github.com/nvidia/warp/commit/9fd571f7b7545c24a9a2d0dabeb2757cecf7b3b3) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=28521896)



> ### Greptile Summary
> This PR adds native, zero-dependency mesh I/O (`wp.load_mesh`, `wp.save_mesh`, `wp.read_mesh`) for OBJ, binary/ASCII STL, and binary/ASCII PLY, cleanly structured behind a `MeshData` dataclass and correctly re-exported into the `wp` namespace. Most P1 issues from prior review rounds have been resolved; one new P1 remains:
> 
> * **CHANGELOG.md**: The [Support `indexedarray` fields in `struct` #1327](https://github.com/NVIDIA/warp/issues/1327) `wp.indexedarray`/`@wp.struct` entry was accidentally deleted — almost certainly a rebase conflict artifact. It must be restored before merge.
> 
> ### Confidence Score: 4/5
> Safe to merge after restoring the accidentally deleted CHANGELOG entry for [GH-1327](https://github.com/NVIDIA/warp/issues/1327); all other remaining findings are P2 quality concerns.
> 
> One P1 remains: an unrelated CHANGELOG entry was removed in the diff (likely a rebase artifact). The parsers and public API are well-structured and the prior P1 bugs (PLY normal allocation, struct format string, STL shape mismatch, negative index off-by-one, MeshData not exported) are all resolved. P2 items (PLY unknown-type misalignment, STL triple-parse, write_stl empty-mesh crash) do not block correctness for the intended use cases.
> 
> CHANGELOG.md (deleted [GH-1327](https://github.com/NVIDIA/warp/issues/1327) entry must be restored)
> 
> ### Important Files Changed
> Filename	Overview
> CHANGELOG.md	New mesh I/O entry added correctly; unrelated [GH-1327](https://github.com/NVIDIA/warp/issues/1327) indexedarray entry was accidentally deleted, likely a rebase conflict artifact.
> warp/_src/io/mesh.py	New dispatch layer: MeshData dataclass, load_mesh/read_mesh/save_mesh/flip_winding helpers — clean structure, correct device handling, well-documented.
> warp/_src/io/obj.py	OBJ reader/writer with fan triangulation, negative index fix, normals length guard; UV array still lacks the same length guard as normals.
> warp/_src/io/ply.py	PLY ASCII/binary reader/writer; binary reader silently misaligns data if any vertex property type is not in the internal type_map.
> warp/_src/io/stl.py	STL binary/ASCII reader with spatial-hash deduplication; ASCII format is reparsed up to three times and write_stl crashes with a confusing error on empty meshes.
> warp/tests/io/test_mesh.py	Comprehensive unittest.TestCase-based suite covering OBJ/STL/PLY load, roundtrip, device, winding-order, and error cases; uses proper Path(**file**) relative paths.
> warp/tests/unittest_suites.py	TestMeshIO and TestMeshIODevices correctly added to default_suite as required by AGENTS.md.
> ### Flowchart
> ```mermaid
> %%{init: {'theme': 'neutral'}}%%
> flowchart TD
>     A["wp.load_mesh(filename, ...)"] --> B["read_mesh(filename, ...)"]
>     C["wp.read_mesh(filename, ...)"] --> B
>     B --> D{File exists?\nSize OK?}
>     D -- No --> E[FileNotFoundError / ValueError]
>     D -- Yes --> F{Detect format}
>     F -- ext=obj --> G["read_obj()"]
>     F -- ext=stl --> H["read_stl()"]
>     F -- ext=ply --> I["read_ply()"]
>     H --> J{Binary or ASCII?}
>     J -- binary --> K["_read_binary_stl()"]
>     J -- ascii --> L["_read_ascii_stl()"]
>     K --> M["_deduplicate_stl_vertices()"]
>     L --> N["_deduplicate_stl_vertices()"]
>     I --> O{Format?}
>     O -- ascii --> P["_read_ply_ascii()"]
>     O -- binary --> Q["_read_ply_binary()"]
>     G --> R[MeshData]
>     M --> R
>     N --> R
>     P --> R
>     Q --> R
>     R --> S{flip_winding?}
>     S -- Yes --> T["_apply_flip_winding()"]
>     S -- No --> U[MeshData returned]
>     T --> U
>     U -- load_mesh path --> V["MeshData.to_warp_mesh(device)"]
>     V --> W[wp.Mesh with BVH]
>     X["wp.save_mesh(mesh, filename)"] --> Y{Detect ext}
>     Y -- obj --> Z["write_obj()"]
>     Y -- stl --> AA["write_stl()"]
>     Y -- ply --> AB["write_ply()"]
> ```
> ### Comments Outside Diff (1)
> 1. `CHANGELOG.md`, line 34-36 ([link](https://github.com/nvidia/warp/blob/9fd571f7b7545c24a9a2d0dabeb2757cecf7b3b3/CHANGELOG.md#L34-L36))
>    [P1](#) **Unrelated changelog entry accidentally removed**
>    The [Support `indexedarray` fields in `struct` #1327](https://github.com/NVIDIA/warp/issues/1327) entry for `wp.indexedarray` support in `@wp.struct` was deleted in this diff. This is almost certainly a rebase/merge-conflict resolution artifact — the AGENTS.md guide specifically warns about this pattern ("After rebasing, diff CHANGELOG.md against the target branch to catch duplicate entries or other issues from clean but incorrect merges"). Restoring the dropped lines is needed to preserve the release notes for that feature.
> 
> Reviews (8): Last reviewed commit: ["fix(io): Discard OBJ UVs when count mism..."](https://github.com/nvidia/warp/commit/9fd571f7b7545c24a9a2d0dabeb2757cecf7b3b3) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=28521896)

One aspect appears to be a false positive. The CHANGELOG diff shows only additions (mesh I/O entry) with no deletions. No GH-1327 entry exists in the main branch CHANGELOG.md to have been accidentally removed.

**Comment by dhairyashilRG** [comment]
I've addressed the initial feedback from the automated bots (fixed the keyword arguments, resolved the OBJ normals shape mismatch, and updated the PLY parsing). All local tests are passing. Ready for a human review whenever someone has time!
